pinc/graph_data.inc

```diff
@@ -650,64 +650,45 @@ function pages_daily($tally_name, $c_or_i, $timeframe)
-        // is correct.
-        if ($start_timestamp == 0) {
-            $where_start_timestamp = $start_timestamp + ($days_to_average * 60 * 60 * 24);
+        // To calculate a simple moving average over $days_to_average we need
+        // to have at least that many days worth of data before $start_timestamp.
+        // This is simple for everything except right at the beginning when
+        // stats were collected. For that we need to jump through some hoops
+        // to 1) detect it and 2) pad the data with 0s so that we can calculate
+        // the SMA and have it align with the real data.
+        $sma_start_timestamp = $start_timestamp - ($days_to_average * 60 * 60 * 24);
+
+        $deltas = $site_tallyboard->get_deltas(1, $sma_start_timestamp, $end_timestamp);
+        $first_timestamp = array_key_first($deltas);
+        $delta_values = array_values($deltas);
+
+        // Determine how much padding, if any, is needed
+        $days_before_first_timestamp = max(($start_timestamp - $first_timestamp) / (60 * 60 * 24), 0);
+        if ($days_before_first_timestamp < $days_to_average) {
+            for ($i = 0; $i <= ($days_to_average - $days_before_first_timestamp); $i++) {
+                array_unshift($delta_values, 0);
+            }
         }
 
-        $sql = "
-            SELECT t1.timestamp,
-                (SELECT round(sum(t2.tally_delta)/count(t2.tally_delta))
-                FROM past_tallies as t2
-                WHERE t2.holder_type = 'S'
-                    AND t2.holder_id = '1'
-                    AND t2.tally_name = '$tally_name'
-                    AND datediff(from_unixtime(t1.timestamp), from_unixtime(t2.timestamp)) between 0 and $days_to_average )
-                as 'sma'
-            FROM past_tallies as t1
-            WHERE t1.holder_type = 'S'
-                AND t1.tally_name = '$tally_name'
-                AND t1.holder_id = '1'
-                AND t1.timestamp > $where_start_timestamp
-            ORDER BY t1.timestamp asc
-        ";
-        $res = DPDatabase::query($sql);
-
-        // Get the earliest start date to use in our population of $moving_average
-        // Because the results are sorted ascending by timestamp, the first one
-        // is the one we want.
-        $earliest_timestamp = null;
-
+        $smas = calculate_sma($delta_values, $days_to_average);
         // store the results in a date-based array we can use to populate the
         // graph's data array
-        while ($result = mysqli_fetch_assoc($res)) {
-            if ($earliest_timestamp === null) {
-                $earliest_timestamp = $result["timestamp"];
+        foreach ($deltas as $timestamp => $delta) {
+            if ($timestamp < $start_timestamp) {
+                continue;
             }
-
-            $average_lookup[date("Y-m-d", $result["timestamp"])] = $result["sma"];
+            $average_lookup[date("Y-m-d", $timestamp)] = array_shift($smas);
```

Comment:
For what it's worth, I suspect this date formatting will be relatively costly to do in bulk. Could we just use `$day = floor($timestamp / 86400);` to calculate (integer) days since epoch instead?

Reply:
Maybe, but 4 calls to this function load in under a second on TEST, which is operating under the same quantity of data in PROD at this point in the script, so it doesn't feel like it's worth the optimization.

Reply:
I think that `floor($timestamp / 86400)` could work for PGDP, because we're far enough from the prime meridian that variations due to DST don't change the GMT-day within which server midnight falls.

I don't think it would work for a DP site whose server time was British Time (with DST). E.g., consider the day they switch to BDST (spring forward, lose an hour): the midnight that starts that server-day occurs at 0:00 GMT, but the midnight that ends that server-day is at 23:00 GMT *of the same GMT day*, so `floor($timestamp / 86400)` would be the same for the two different midnights. Conversely, when they fall back (gain an hour), there'd be a GMT-day that appeared to be skipped by that formula.